httpcore/_async/http_proxy.py

```diff
@@ -71,15 +65,11 @@ async def request(
         if self._keepalive_expiry is not None:
             await self._keepalive_sweep()
 
-        if (
-            self.proxy_mode == "DEFAULT" and url[0] == b"http"
-        ) or self.proxy_mode == "FORWARD_ONLY":
```

Comment:
So, I don't think that *this pull request* should change this behaviour. (Or at least if it *should*, then I haven't understood why it needs to do so.) I think we *should* have a discussion about the `"DEFAULT"/"FORWARD_ONLY"/"TUNNEL_ONLY"`, but I *don't* think we need to conflate that with the rest of this PR right?

Is there any reason why we shouldn't...

* Keep the API exactly as it is, in this pull request.
* Get the pull request merged.
* Treat the `proxy_mode` API discussion independently to this fix?

Reply:
We can do all of those things, of course. 🙂

What's key here is the definition of "keeping the API as it is":

If we mean strictly the possible values of `proxy_mode` then nothing's broken, we still `DEFAULT`, `FORWARD_ONLY` or `TUNNEL_ONLY`.

However, this PR introduces a _behaviour_ change, where `DEFAULT` now means `TUNNEL_ONLY` _always_ and also requires the caller to sent the appropriate `Host` header. That logic was removed as part of a review.

I'm happy to reintroduce that logic if we feel we're breaking the contract with httpx, OTOH we can add it there and keep them in sync. I don't mind either way.

Reply:
> this PR introduces a *behaviour* change

Sure. I'd suggest that we don't introduce that change here, and instead consider it in isolation to the rest of this PR, since I don't *think* it's strictly required for the rest of this right?

I can see an argument in favour of making that change too, but it's complex because there's a knock-on effect w/ how it'd affect the `httpx` API & implementation, that I think'd be worth keeping as separately reviewable piece of work.

Reply:
No probs, I'll add it back soon. Thanks!

Reply:
I reinstated the original `DEFAULT` behaviour. I actually misspoke before:

> also requires the caller to sent the appropriate Host header. That logic was removed as part of a review.

That logic was never part of httpcore, in 0.12.1 `httpx` used a `Request` object for the `CONNECT` request which automatically added the `host` header. I thought we did it here because in the first iteration of this PR I [added it](https://github.com/encode/httpcore/pull/57#discussion_r407209432).

As it stands `httpcore` requires the user to set the appropriate proxy headers.